Send preload links via HTTP Link headers in addition to LINK tags

AhmarZaidi opened the pull request:
## Summary

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Fixes #1321

## Relevant technical choices

<!-- Please describe your changes. -->

- Added a new method `get_headers()` to `OD_Preload_Link_Collection` to construct the Link HTTP response headers for preloading resources.
- Updated `od_optimize_template_output_buffer` to call `headers()` if there are preload links available and headers have not been sent already.
- Sent HTTP Link headers separately (above appending the link tags) to prevent [`Cannot modify header information - headers already sent`](https://stackoverflow.com/a/8028987/19726929) error due to output before sending headers.

<!--
For maintainers only, please make sure:

- PR has a `[Type]` label.
- PR has a plugin-specific milestone, or the `no milestone` label if it does not apply to any specific plugin.
- PR has a changelog-friendly title, or the `skip changelog` label if it should not be mentioned in the plugin's changelog.
-->

AhmarZaidi committed:
Add support for sending preload links via HTTP headers

AhmarZaidi commented:
Please add type label and milestone.

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +166,37 @@ public function get_html(): string {
+	 *
+	 * @return string HTTP Link header.
+	 */
+	public function get_headers(): string {
```
Let's call this `get_response_header`

westonruter replied:
Let's call this `get_response_header`

westonruter replied:
```suggestion
	public function get_response_header(): ?string {
```

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +166,37 @@ public function get_html(): string {
 	}
 
+	/**
+	 * Gets the HTTP Link header string.
```
```suggestion
	 * Constructs the Link HTTP response header.
```

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +166,37 @@ public function get_html(): string {
+	/**
+	 * Gets the HTTP Link header string.
+	 *
+	 * @return string HTTP Link header.
```
```suggestion
	 * @return string|null Link HTTP response header, or null if there are none.
```

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +166,37 @@ public function get_html(): string {
+			}
+			$link['attributes']['media'] = implode( ' and ', $media_features );
+
+			$link_header = '<' . esc_url( $link['attributes']['href'] ?? '' ) . '>; rel="preload"';
```
The `esc_url()` function includes HTML entities. So this should rather be:

```suggestion
			$link_header = '<' . esc_url_raw( $link['attributes']['href'] ?? '' ) . '>; rel="preload"';
```

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +166,37 @@ public function get_html(): string {
+
+			$link_headers[] = $link_header;
+		}
+
```
```suggestion
		if ( count( $link_headers ) === 0 ) {
				return null;
		}

```

westonruter commented on plugins/optimization-detective/optimization.php:
```diff
@@ -164,6 +164,11 @@ function od_optimize_template_output_buffer( string $buffer ): string {
 
 	$tag_visitor_registry = new OD_Tag_Visitor_Registry();
 
+	// Send any preload links as Link headers.
+	if ( count( $preload_links ) > 0 && ! headers_sent() ) {
+		header( $preload_links->get_headers() );
+	}
+
```
Let's move this down to combine with the existing condition:

```php
	// Inject any preload links at the end of the HEAD.
	if ( count( $preload_links ) > 0 ) {
		$walker->append_head_html( $preload_links->get_html() );
	}
```

For example:

```php
// Send any preload links in a Link response header and in a LINK tag injected at the end of the HEAD.
if ( count( $preload_links ) > 0 ) {
	header( $preload_links->get_response_header() );
	$walker->append_head_html( $preload_links->get_html() );
}
```

AhmarZaidi replied:
Hey, so I tried this but encountered errors while running tests.
```bash
➜  wordpress-performance git:(preload-links) npm run test-php
...
There were 9 errors:

1) Test_Image_Prioritizer_Helper::test_image_prioritizer_register_tag_visitors with data set "common-lcp-image-and-lazy-loaded-image-outside-viewport-with-fully-populated-sample-data" (Closure Object (...), '\n                                      <html lang="en">\n                              ...\n                           ', '\n <html lang="en">\n                               ...\n                           ')
Cannot modify header information - headers already sent by (output started at /wordpress-phpunit/includes/bootstrap.php:265)

2) Test_Image_Prioritizer_Helper::test_image_prioritizer_register_tag_visitors with data set "common-lcp-image-with-fully-incomplete-sample-data" (Closure Object (...), '\n                                    <html lang="en">\n                              ...\n                           ', '\n                                  <html lang="en">\n                              ...\n                           ')
Cannot modify header information - headers already sent by (output started at /wordpress-phpunit/includes/bootstrap.php:265)
...
``` 

This is the line in `/wordpress-phpunit/includes/bootstrap.php:265` referenced by the error where the output is started:
```
system( WP_PHP_BINARY . ' ' . escapeshellarg( __DIR__ . '/install.php' ) . ' ' . escapeshellarg( $config_file_path ) . ' ' . $ms_tests . ' ' . $core_tests, $retval );
```

I believe [this section](https://github.com/WordPress/performance/blob/trunk/plugins/optimization-detective/optimization.php#L176-L190) of the `plugins/optimization-detective/optimization.php` is producing some output but `headers()` should be called before any output because after starting to write the http content, headers cannot be sent in the http response as mentioned [here](https://stackoverflow.com/a/8028987/19726929).

Can we move the `$walker->append_head_html( $preload_links->get_html() );` to the above section?
Or is there another solution to this?

Also, should I remove the check for `headers_sent()`?

westonruter replied:
Oh yes, keep the `headers_sent()` check. The issue is that PHPUnit itself is printing output, so it is preventing headers from being sent.

westonruter replied:
So like:

```php
// Send any preload links in a Link response header and in a LINK tag injected at the end of the HEAD.
if ( count( $preload_links ) > 0 ) {
	if ( ! headers_sent() ) {
		header( $preload_links->get_response_header() );
	}
	$walker->append_head_html( $preload_links->get_html() );
}
```

westonruter replied:
Oh, one more important thing: make sure to pass `false` as the second argument to the `header()` function. This will prevent it from overriding any existing `Link` headers that may have been sent.

AhmarZaidi replied:
Thanks, will implement these changes.

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +166,37 @@ public function get_html(): string {
+	public function get_headers(): string {
+		$link_headers = array();
+
+		foreach ( $this->get_adjacent_deduplicated_links() as $link ) {
```
This foreach has quite a bit of duplication with `get_html()`. Can the common code be factored out into a private helper method?

AhmarZaidi replied:
Sure makes sense, will do that

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +166,37 @@ public function get_html(): string {
+			$link_header = '<' . esc_url( $link['attributes']['href'] ?? '' ) . '>; rel="preload"';
+			foreach ( $link['attributes'] as $name => $value ) {
+				if ( 'href' !== $name ) {
+					$link_header .= sprintf( '; %s="%s"', $name, esc_attr( $value ) );
```
Use of `esc_attr()` isn't right here since this is not an HTML context.  I believe you should instead percent-encode the value here using `rawurlencode()`.

westonruter reviewed (changes requested):
Thank you! Great start.

westonruter commented:
Oh, we'll also need to add tests for this new method.

AhmarZaidi committed:
PR Changes

- Remove code duplication
- Update escaping functions
- Update documentation
- Handle no headers case

AhmarZaidi committed:
Merge branch 'trunk' into preload-links

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -136,6 +136,27 @@ static function ( array $carry, array $link ): array {
+	/**
+	 * Adds media features to the links.
+	 *
+	 * @phpstan-param Link $link
+	 * @param array $link Link array.
+	 * @phpstan-return Link
+	 * @return array Link array with media features added.
```
```suggestion
	 * @param Link $link Link array.
	 * @return Link Array with media features added.
```

westonruter replied:
Nevermind, I guess these are needed.

AhmarZaidi replied:
Yes, I was looking for a better way but kept getting either phpstan or linter error without this.

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -145,14 +166,7 @@ public function get_html(): string {
-				$media_features[] = sprintf( '(max-width: %dpx)', $link['maximum_viewport_width'] );
-			}
-			$link['attributes']['media'] = implode( ' and ', $media_features );
+			$link = $this->add_media_features( (array) $link );
```
Is the casting necessary?

```suggestion
			$link = $this->add_media_features( $link );
```

westonruter replied:
Alternatively, if the above suggestion is accepted, then:

```suggestion
			$link['media] = $this->get_media_attribute( $link );
```

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -136,6 +136,27 @@ static function ( array $carry, array $link ): array {
+	 * @phpstan-return Link
+	 * @return array Link array with media features added.
+	 */
+	private function add_media_features( array $link ): array {
+		$media_features = array( 'screen' );
+		if ( null !== $link['minimum_viewport_width'] && $link['minimum_viewport_width'] > 0 ) {
+			$media_features[] = sprintf( '(min-width: %dpx)', $link['minimum_viewport_width'] );
+		}
+		if ( null !== $link['maximum_viewport_width'] && PHP_INT_MAX !== $link['maximum_viewport_width'] ) {
+			$media_features[] = sprintf( '(max-width: %dpx)', $link['maximum_viewport_width'] );
+		}
+		$link['attributes']['media'] = implode( ' and ', $media_features );
+
+		return (array) $link;
+	}
```
However, instead of amending the link, how about just returning the media features string instead?

```suggestion
	private function get_media_attribute( array $link ): string {
		$media_features = array( 'screen' );
		if ( null !== $link['minimum_viewport_width'] && $link['minimum_viewport_width'] > 0 ) {
			$media_features[] = sprintf( '(min-width: %dpx)', $link['minimum_viewport_width'] );
		}
		if ( null !== $link['maximum_viewport_width'] && PHP_INT_MAX !== $link['maximum_viewport_width'] ) {
			$media_features[] = sprintf( '(max-width: %dpx)', $link['maximum_viewport_width'] );
		}
		return implode( ' and ', $media_features );
	}
```

westonruter replied:
Further still, what if this logic were rather incorporated into `get_adjacent_deduplicated_links()`? Instead of returning the result of the `array_reduce()` it could then loop over the links and construct the `media` attribute for each. This method could then be changed from `get_adjacent_deduplicated_links()` to something like `prepare_links()`.

AhmarZaidi replied:
This approach seems better. Will update the `get_adjacent_deduplicated_links()` function.

AhmarZaidi committed:
Prepare links

- Deduplicate adjacent links
- Add media attributes

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -76,14 +76,14 @@ public function add_link( array $attributes, ?int $minimum_viewport_width, ?int
 
 	/**
-	 * Get adjacent-deduplicated links.
+	 * Prepare links by deduplicating adjacent links and adding media attributes.
```
```suggestion
	 * Prepares links by deduplicating adjacent links and adding media attributes.
```

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -76,14 +76,14 @@ public function add_link( array $attributes, ?int $minimum_viewport_width, ?int
+	 * @return array<int, Link> Prepared links with adjacent-duplicates merged together and media attributes added.
 	 */
-	private function get_adjacent_deduplicated_links(): array {
+	private function prepare_links(): array {
```
I know I suggested `prepare_links()` but could we go with `get_prepared_links()` instead? I think that would be a _tiny_ bit better.

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -100,7 +100,8 @@ static function ( array $a, array $b ): int {
 		);
 
-		return array_reduce(
+		// Deduplicating adjacent links.
```
```suggestion
		// Deduplicate adjacent links.
```

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +172,31 @@ public function get_html(): string {
+		$link_headers = array();
+
+		foreach ( $this->prepare_links() as $link ) {
+			$link_header = '<' . esc_url_raw( $link['attributes']['href'] ?? '' ) . '>; rel="preload"';
```
Is it valid to send a `Link` without the header? It seems that the [uri-reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link#uri-reference) is required. In reality, the only reason why `href` was made optional was for the sake of preloading responsive images since Safari at the time didn't support `imagesrcset` and `imagesizes`, so adding a preload link that omitted the `href` was a way to prevent Safari from needlessly preloading an image which might not be used. 

Nevertheless, as seen in [this web.dev article](https://web.dev/articles/preload-responsive-images#image-set:~:text=Responsive%20image%20preloading%20lets%20you%20load%20those%20images%20faster.), there is still a use case for a preload link without `href` and that is for preloading background images which have different pixel densities:

```html
<link rel="preload" as="image" imagesrcset="cat.png 1x, cat-2x.png 2x">
```

So it seems we do need `href` to remain able to be optional.

What to do in the case where there is no `href`?  I just made a [test glitch](https://imagesrcset-only-preload-link-http-header.glitch.me/) to check various scenarios. You can see that [when an empty string is used](https://imagesrcset-only-preload-link-http-header.glitch.me/?preload_url_type=empty-string), there is nothing downloaded:

![image](https://github.com/WordPress/performance/assets/134745/6174a595-b160-45a2-a497-0f3b7741e536)

In contrast, check what happens [when `about:blank` is used as the URL](https://imagesrcset-only-preload-link-http-header.glitch.me/?preload_url_type=about-blank):

![image](https://github.com/WordPress/performance/assets/134745/581b8b08-a0f5-40f4-8c91-77f5ae3f88f2)

It works!

So let's do this:

```suggestion
			// The about:blank is present since a Link without a reference-uri is invalid so any imagesrcset would otherwise not get downloaded.
			$link_header = '<' . ( isset( $link['attributes']['href'] ) ? esc_url_raw( $link['attributes']['href'] ) : 'about:blank' ) . '>; rel="preload"';
```

westonruter replied:
I also checked and using `about:blank` here works in Firefox and Safari too. This doesn't seem to be part of the spec, so it could be an implementation accident. But the alternative is to skip sending the `Link` altogether. So I think using `about:blank` is best.

AhmarZaidi replied:
I see, thanks this was very insightful.

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +172,31 @@ public function get_html(): string {
+
+		foreach ( $this->prepare_links() as $link ) {
+			$link_header = '<' . esc_url_raw( $link['attributes']['href'] ?? '' ) . '>; rel="preload"';
+			foreach ( $link['attributes'] as $name => $value ) {
+				if ( 'href' !== $name ) {
+					$link_header .= sprintf( '; %s="%s"', $name, rawurlencode( $value ) );
+				}
```
```suggestion
			unset( $link['attributes] );
			foreach ( $link['attributes'] as $name => $value ) {
				$link_header .= sprintf( '; %s="%s"', $name, rawurlencode( $value ) );
```

AhmarZaidi replied:
Shouldn't we add a check if `$link['attributes']['href']` is set before unsetting?
```php
if ( isset( $link['attributes']['href'] ) ) {
    unset( $link['attributes']['href'] );
}
foreach ( $link['attributes'] as $name => $value ) {
    $link_header .= sprintf( '; %s="%s"', $name, rawurlencode( $value ) );
}
```

AhmarZaidi replied:
How about something like this:
```php
$link['attributes']['href'] = isset( $link['attributes']['href'] ) ? esc_url_raw( $link['attributes']['href'] ) : 'about:blank';
$link_header = '<' . $link['attributes']['href'] . '>; rel="preload"';
unset( $link['attributes']['href'] );
foreach ( $link['attributes'] as $name => $value ) {
    $link_header .= sprintf( '; %s="%s"', $name, rawurlencode( $value ) );
}
```

westonruter replied:
Yes, that looks good.

westonruter reviewed (changes requested):
Getting close!

AhmarZaidi commented:
> Oh, we'll also need to add tests for this new method.

Had some doubts regarding tests implementation:

1. I believe [`get_html()`](https://github.com/WordPress/performance/blob/trunk/plugins/optimization-detective/class-od-preload-link-collection.php#L144) is tested by using the output of [`od_optimize_template_output_buffer()`](https://github.com/WordPress/performance/blob/1ed2fbdab5fa47e2b174b813f6d5d3273b589482/plugins/optimization-detective/optimization.php#L143) since the walker adds links to the markup and compares it with the data from data provider.
Now, how would I test for `get_response_header()` since they are added to the http response and not the markup.
I explored `xdebug_get_headers()` but that doesn't seem to work. Could separately test `get_response_header()` function's output but I don't think that's the right way.

2. If we implement it like this in `optimization.php`: 
> ```php
> // Send any preload links in a Link response header and in a LINK tag injected at the end of the HEAD.
> if ( count( $preload_links ) > 0 ) {
> 	if ( ! headers_sent() ) {
> 		header( $preload_links->get_response_header() );
> 	}
> 	$walker->append_head_html( $preload_links->get_html() );
> }
> ```
Then `headers()` won't be called while testing because `headers_sent()` would be true due to [prior output](https://github.com/WordPress/performance/pull/1323#discussion_r1661437951), so how would we test it?

AhmarZaidi committed:
PR Changes

- Update documentation
- Fix preload link without href using about:blank fallback

westonruter commented:
@AhmarZaidi Yeah, testing the output of `header()` is a pain. I don't think we can practically test it being output when calling `od_optimize_template_output_buffer()`. I think it's fine to just test that `get_response_header()` method in isolation.

AhmarZaidi committed:
Add test for http link response headers

westonruter commented on plugins/optimization-detective/tests/test-optimization.php:
```diff
@@ -368,6 +368,44 @@ function ( OD_HTML_Tag_Walker $walker, OD_URL_Metrics_Group_Collection $url_metr
+	/**
+	 * Test get_response_header().
+	 *
+	 * @covers \OD_Preload_Link_Collection::get_response_header
```
```suggestion
	 * @covers OD_Preload_Link_Collection::get_response_header
```

westonruter commented on plugins/optimization-detective/tests/test-optimization.php:
```diff
@@ -368,6 +368,44 @@ function ( OD_HTML_Tag_Walker $walker, OD_URL_Metrics_Group_Collection $url_metr
+			1200
+		);
+
+		$expected_header = 'Link: <https://example.com/foo.jpg>; rel="preload"; as="image"; fetchpriority="high"; imagesrcset="https%3A%2F%2Fexample.com%2Ffoo-480w.jpg%20480w%2C%20https%3A%2F%2Fexample.com%2Ffoo-800w.jpg%20800w"; imagesizes="%28max-width%3A%20600px%29%20480px%2C%20800px"; crossorigin="anonymous"; media="screen", <https://example.com/bar.jpg>; rel="preload"; as="image"; fetchpriority="h ...
```
The `imagesrcset`, `imagesizes`, and `media` attribute values here seem wrong being URL-encoded. Do these work being encoded like this?

westonruter replied:
Indeed. I found it doesn't work, for `imagesrcset` at least. I tried sending that `Link` header on a Glitch and I saw a network request attempted for:

```
https://alert-sustaining-acrylic.glitch.me/https%3A%2F%2Fexample.com%2Ffoo-480w.jpg%20480w%2C%20https%3A%2F%2Fexample.com%2Ffoo-800w.jpg%20800w
```

westonruter replied:
I think I got this. See d1cd8d4

westonruter commented on plugins/optimization-detective/class-od-preload-link-collection.php:
```diff
@@ -166,6 +172,32 @@ public function get_html(): string {
+			$link_header                = '<' . $link['attributes']['href'] . '>; rel="preload"';
+			unset( $link['attributes']['href'] );
+			foreach ( $link['attributes'] as $name => $value ) {
+				$link_header .= sprintf( '; %s="%s"', $name, rawurlencode( $value ) );
```
Per below, we may need to reconsider the use of `rawurlencode()` here. Maybe it's just that the generated header test "looks" bad.

westonruter committed:
Remove unnecessary global namespace backslash

westonruter committed:
Escape quoted strings according to RFC 9110

westonruter committed:
Merge branch 'trunk' into preload-links

westonruter committed:
Use array_map() instead of reference

westonruter commented:
Great work!
